samples/extensions/khr/externalmemory/vulkan_utils.h

```diff
@@ -0,0 +1,382 @@
+                    compatible_dev_index = cl_candidates_count;
+                    break;
+                }
+            }
+        }
+
+        if (compatible_dev_index == cl_candidates_count)
+        {
+            return false;
+        }
+
+        candidate->vk_props = props2.properties;
+        candidate->cl_candidate = cl_candidates[compatible_dev_index];
+    }
+
+    // Check if the device supports the required extensions.
+    if (!check_device_extensions(vk_device, required_device_extensions,
+                                 required_device_extensions_count))
+    {
+        return false;
+    }
+
+    candidate->vk_candidate = vk_device;
+    return true;
+}
+
+// Check if a given OpenCL device supports a particular set of Khronos
+// extensions.
+bool check_khronos_extensions(
+    const cl_device_id cl_device,
+    const char* const* const required_khronos_extensions,
+    const size_t required_khronos_extensions_count)
+{
+    cl_int error = CL_SUCCESS;
+    size_t supported_extensions_count;
+    OCLERROR_RET(clGetDeviceInfo(cl_device, CL_DEVICE_EXTENSIONS, 0, NULL,
+                                 &supported_extensions_count),
+                 error, ret);
+    char* supported_extensions =
+        (char*)malloc(supported_extensions_count * sizeof(char));
+    OCLERROR_RET(clGetDeviceInfo(cl_device, CL_DEVICE_EXTENSIONS,
+                                 supported_extensions_count,
+                                 supported_extensions, NULL),
+                 error, err);
+
+    for (size_t i = 0; i < required_khronos_extensions_count; ++i)
+    {
+        if (!strstr(supported_extensions, required_khronos_extensions[i]))
+        {
+            free(supported_extensions);
+            return false;
+        }
+    }
+    free(supported_extensions);
+    return true;
+err:
+    free(supported_extensions);
+ret:
+    return false;
+}
+
+// Find a suitable device for the example, that is, an OpenCL
+// device that is also Vulkan-compatible and that supports the required
+// Vulkan device extensions.
+struct device_candidate
+find_suitable_device(VkInstance instance,
+                     const char* const* required_device_extensions,
+                     const size_t required_device_extensions_count)
+{
+    // Query OpenCL devices available.
+    cl_int error = CL_SUCCESS;
+    bool candidate_found = false;
+    cl_uint cl_platform_count = 0;
+    struct device_candidate found_candidate = {0};
+    OCLERROR_RET(clGetPlatformIDs(0, NULL, &cl_platform_count), error, ret);
+
+    cl_platform_id* platforms =
+        (cl_platform_id*)malloc(cl_platform_count * sizeof(cl_platform_id));
+    OCLERROR_RET(clGetPlatformIDs(cl_platform_count, platforms, NULL), error,
+                 platforms);
+
+    size_t cl_device_count = 0;
+    const char* uuid_khronos_extension[] = {
+        CL_KHR_DEVICE_UUID_EXTENSION_NAME
+    };
+    for (cl_uint cl_platform_id = 0; cl_platform_id < cl_platform_count;
+         ++cl_platform_id)
```

Comment:
It'd be best not to name this loop variable `cl_platform_id` since that is also the name of an OpenCL type.

Reply:
Changed!